qualang_tools/bakery/readme.md

```diff
@@ -0,0 +1,273 @@
+
+    - “symmetric_r' pads 0s symmetrically before and after the baked sample , putting one more 0 after it in case the baked sample length is odd
+
+Declaration is done before the QUA program as follows: 
+
+```
+with baking(config, padding_method = "symmetric_r") as b:
+  b.align("qe1", "qe2", "qe3")
+  b.frame_rotation(0.78, "qe2")
+  b.ramp(amp=0.3, duration=9, qe="qe1")
+```
+
+This context manager does not return any output, its execution results in an edition of the configuration file provided as input to add:
+- an operation for each quantum element involved within the baking context manager
+- an associated pulse
+- an associated waveform (set of 2 waveforms for a mixedInputs quantum element) containing  sample(s) issued from concatenation of operations indicated in the context manager.
+  
+The baking also carries advantages for an efficient program memory management,
+in the sense that there are less commands and different samples to be stored in memory to execute the program.
+
+# **How can I add operations inside the baking context manager?**
+
+The logic behind the baking context manager is to stay as close as possible to the way we would write play statements within a QUA program. For instance, commands like frame_rotation, reset_phase, ramp, wait and align are all replicated  within the context manager. 
+
+For playing custom operations, from an arbitrary shaped waveform the procedure goes as follows:
+
+1. You first have to write down the sample you want to use as a waveform (with arbitrary length, no matter if it does not match usual QUA criteria for saving a waveform in memory) in the form of a Python list.
+    - If the sample is meant for a singleInput element, the list should contain the sample itself. 
+    - Contrariwise, if it is intended for a mixInputs element, the list should contain two Python lists as ```[sample_I, sample_Q]``` , where sample_I and sample_Q are themselves Python lists containing the samples.
+
+2. Add the sample to the local configuration, with method ```add_Op```, which takes 4 inputs: 
+    - The name of the operation (name you will use only within the baking context manager in a play statement)
+    - The quantum element for which you want to add the operation
+    - The samples to store as waveforms
+    - The digital_marker name (supposedly already existing in the configuration) to attach to the pulse associated to the operation.
+
+3. Use a baking ``play()`` statement, specifying the operation name (which should correspond to the name introduced in the add_Op method) and the quantum element to play the operation on
+
+All those commands concatenated altogether eventually build one single “big” waveform per quantum element involved in the baking that contains all the instructions specified in the baking environment. The exiting procedure of the baking ensures that the appropriate padding is done to ensure that QUA will be able to play this arbitrary waveform.

```

Comment:
The exiting procedure of the baking ensures that the appropriate padding is done to ensure that **the OPX** will be able to play this arbitrary waveform.